### ISSUE TITLE
Plugin API: proper stream interface and a method for opening data stream

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#2226]
**Problem**

Engine's Plugin API lacks any means to open game data stream. There are IAGSEngine->FRead and FWrite functions, they require file "handle", an arbitrary int32 value provided by the engine, but in practice these are only usable when saving or reading a game save.

Because of this, if plugin wants to load any custom data, it will have to either use standard FILE* stream, or implement its own reader. This is bothersome, and also not always suitable (some ports may have their own restrictions to how data is packaged or which system API may be used).

**Solution**

The most minimal solution would be to add some kind of IAGSEngine->FOpen function that would resolve a script path with location tokens, open a data stream, and return stream "handle". But after bit of thinking, this may not be enough, because operations in plugin may require few other things such as knowing stream's length and seeking.

Therefore I propose to add a new interface object, with minimal necessary functions for stream use:

```
class IAGSStream {
    virtual int Read(buffer, len);
    virtual int Write(buffer, len);
    virtual int Length();
    virtual int Position();
    virtual void Seek(int offset, int base);
};
```

And two functions in IAGSEngine that open and close it:
```
    IAGSStream* OpenStream(path, mode);
    CloseStream(IAGSStream*);
```

Open function would need to support everything that engine supports internally, that is: opening files on disk, assets in game package (for reading), and stream from any platform-specific packages (e.g. AAssetStream on Android).

**Implementation**

Internally IAGSStream would just wrap our Stream object pointer, hence supporting any stream source.

**Additional notes**

Would need to decide what to do with FRead, FWrite and game saving process in the context of this new interface.
Of course old functions need to keep working for backwards compatibility.

**Comment by ivan-mogilko** [comment]
I am planning to make a draft of this and test with the "spritevideo" plugin I'm currently (struggling) to port to Android.
https://github.com/ivan-mogilko/ags-spritevideo/tree/android

**Comment by ivan-mogilko** [comment]
Resolved by #2227.
